In [15]:
import pandas as pd
import numpy as np
import matplotlib as plt
import plotly.express as px


In [154]:
data = pd.read_csv('C:/Users/фора/Desktop/dataset.csv')
data.head()
data=data.drop(index=[0])

In [159]:
# 1. What is the average age and percentage of road accidents for all drivers based on genders? 
data_wac=data[data['Дата проишествия'].notnull()]

data_wac=data_wac['Пол'].value_counts().reset_index()


fig = px.pie(data_wac['index'], values=data_wac['Пол'], names=data_wac['index'])
fig.update_traces( textinfo='percent', textfont_size=15)
fig.update_layout(title="distribution of road accidents between two genders",title_x=0.5)
fig.show()

In [160]:
#2. What is the difference between average driving experience and mean bonus-malus class for males and females? 
data_male=data[data['Пол']=='М']
male=data_male['Стаж вождения'].mean() - data_male['КБМ'].mean()
data_female=data[data['Пол']=='Ж']
female=data_female['Стаж вождения'].mean() - data_female['КБМ'].mean()
print('Difference between values in male:',male)
print('Difference between values in female:',female)


Difference between values in male: 11.302363399826538
Difference between values in female: 7.099703264094955


In [162]:
#3. In which year the most significant number of vehicles were built?

data['Год'] = data['Год выпуска'].str.slice(-4)
data['Год'].value_counts().head(1)

2007    715
Name: Год, dtype: int64

**2007 year with 715 values**

In [163]:
#3.2. Please provide the list of top-10 years 
data['Год'].value_counts().head(10)

2007    715
1995    653
1996    594
1993    594
1992    561
1994    520
2002    498
1997    497
1991    489
2006    467
Name: Год, dtype: int64

In [212]:
# 4. How much customers pay for the contract in average and how this number is correlated with the bonusmalus class?
average_pay=data['Страховая премия'].mean()
print(average_pay)


s=data[['КБМ', 'Страховая премия']]
s.corr() 

# there is weak correlation coefficent , so they are not related. 


10931.720139334495


КБМ  Страховая премия
КБМ               1.000000         -0.053462
Страховая премия -0.053462          1.000000

In [164]:
# 5. What is the most popular vehicle-make company in any 5 cities? (choose the cities you like most)
pop=data[(data['Город']=='Шымкент')|(data['Город']=='Нур-Султан') | (data['Город']=='Алматы')| (data['Город']=='Павлодар')| (data['Город']=='Атбасар')]
pop_model=pd.crosstab(pop['Модель'],pop['Город'])
print('The popular car in Almaty',pop_model['Алматы'].sort_values(ascending=False).head(1))
print('The popular car in Shymkent',pop_model['Шымкент'].sort_values(ascending=False).head(1))
print('The popular car in Nur-sultan',pop_model['Нур-Султан'].sort_values(ascending=False).head(1))
print('The popular car in Pavlodar',pop_model['Павлодар'].sort_values(ascending=False).head(1))
print('The popular car in Atbasar',pop_model['Атбасар'].sort_values(ascending=False).head(1)) 

The popular car in Almaty Модель
Toyota    1201
Name: Алматы, dtype: int64
The popular car in Shymkent Модель
Volkswagen    6
Name: Шымкент, dtype: int64
The popular car in Nur-sultan Модель
Toyota    259
Name: Нур-Султан, dtype: int64
The popular car in Pavlodar Модель
Toyota    36
Name: Павлодар, dtype: int64
The popular car in Atbasar Модель
ВАЗ    78
Name: Атбасар, dtype: int64


The popular car in Almaty Toyota    1201

The popular car in Shymkent Volkswagen    6

The popular car in Nur-sultan Toyota    259

The popular car in Pavlodar Toyota    36

The popular car in Atbasar ВАЗ    78


In [186]:
# 6. What kind of vehicle colors people prefer most, dark ones or light ones? Group all colors to dark and light as you like.

#probable distribution of colors 

dark = ['темно','черный','темно-коричневый','темно-коричневый','серый с зеленым отливом металлик', 
        'синий металлик', 'темно-вишневый'] 
light = ['белый','бело','светло','ярко','бежевый','светло-серебристый металлик',
         'золотое руно', 'светло-зелёный', 'серый металлик' ]

data_light=data[data['Цвет'].isin(light)]
data_dark=data[data['Цвет'].isin(dark)]

print('number of light colored cars :',data_light.shape[0])
print('number of dark colored cars :',data_dark.shape[0])


number of light colored cars : 2470
number of dark colored cars : 1216


Light colored cars are more than dark colored

In [53]:
# 7. What is the most popular vehicle-make company for females in Almaty, Shymkent and Nur-Sultan (Astana) respectively?
data_female_new=data_female[(data_female['Город']=='Шымкент')|(data_female['Город']=='Нур-Султан') | (data_female['Город']=='Алматы')]
pop_among_women=pd.crosstab(data_female_new['Модель'],data_female_new['Город'])
print(pop_among_women['Нур-Султан'].sort_values(ascending=False).head(1))
print(pop_among_women['Шымкент'].sort_values(ascending=False).head(1))
print(pop_among_women['Алматы'].sort_values(ascending=False).head(1))



Модель
Toyota    58
Name: Нур-Султан, dtype: int64
Модель
Audi    1
Name: Шымкент, dtype: int64
Модель
Toyota    297
Name: Алматы, dtype: int64


Nur-Sultan : Toyota    58

Shymkent : Audi    1

Almaty : Toyota    297

**Number next to model of car is number of occurence**

In [209]:
# 8.1 What is the distribution of car road accidents for different cities? 
road_accident=data[data['Дата проишествия'].notnull()]
road_accidents=road_accident['Город'].value_counts().reset_index()
road_accidents

fig = px.bar(road_accidents, x=road_accidents['index'], y=road_accidents['Город'],
              color='index',
             labels={'index':'number of accidents between cities'}, height=400)
fig.show()

In [188]:
# 8.2 Which city is in the first place for road accidents?
road_accident['Город'].value_counts().head(1)
#Almaty is in the first place by car accidents

Алматы    56
Name: Город, dtype: int64

Almaty is in top with 56 accidents

In [187]:
# 8.3 From the top-10 cities, is there any countrysides?
list_of_cities=road_accident['Город'].value_counts().reset_index()
top_cities=['Алматы','Нур-Султан','Шымкент','Актобе','Караганда','Тараз','Павлодар','Атырау','Усть-Каменогорск','Семей']
# creating top_cities list by top-10 and extracting python list comprehension by isin function is equaled false
list_of_cities[list_of_cities['index'].isin(top_cities)==False]

index  Город
1     Петропавловск     17
4          Костанай     10
5          Кокшетау      5
8           Атбасар      2
9   Атбасар,Атбасар      2
11         Житикара      1
15    Отеген батыра      1
16           Талгар      1
17        Экибастуз      1
18        Аксу,Аксу      1

In [214]:
# 9. How many people had more than 1 car in one year? Is there any driver who had more than 1 car accident in one year?
data_ppl = data.loc[:, ["Unique number", "Марка", "Модель", "Дата проишествия"]]
data_ppl["Марка и модель"] = data_ppl["Марка"] + " " + data_ppl["Модель"]
data_ppl = data_ppl.drop(columns=['Модель'])
data_ppl["Дата проишествия"] = pd.DatetimeIndex(data_ppl["Дата проишествия"]).year
data_ppl = data_ppl.groupby(by=["Unique number"]).count().sort_values("Дата проишествия", ascending=False)
data_ppl = data_ppl.reset_index()
data_ppl[(data_ppl["Марка и модель"]>1) & (data_ppl["Дата проишествия"]>1)]
#more than 1 accident and more than one car


Unique number  Марка  Дата проишествия  Марка и модель
0  0a3a1e71a726f4744afbdb2ef1357148     13                 7              13
1  1bbce6dfdba970d99f41ed1f44488fc8     12                 3              12
2  14f69489b14639e440988e3a52a650e5      2                 2               2
3  d2f959644b31e72df511f402d1136141      2                 2               2
4  56a4c2beb2e15a034bb3c2fc0c544b7e      2                 2               2

In [108]:
# 10. What is the distribution of road accidents based on the bonus-malus class?
distribution=data.value_counts('КБМ')
import plotly.express as px

distribution=distribution.reset_index()
distribution=distribution.rename(columns={0: 'appearance'})

fig = px.pie(distribution['КБМ'], values=distribution['appearance'], names=distribution['КБМ'])
fig.update_traces( textinfo='percent', textfont_size=15)
fig.update_layout(title="Bonus-malus class distribution",title_x=0.5)
fig.show()
# distribution